src/dynamic_query.rs

```diff
@@ -0,0 +1,130 @@
+use core::{alloc::Layout, any::TypeId, ptr::NonNull};
+
+use crate::{entities::EntityMeta, Archetype, Entity};
+
+/// The component types accessed by a dynamic query.
+#[derive(Debug, Copy, Clone)]
+pub struct DynamicQueryTypes<'a> {
+    pub(crate) read_types: &'a [TypeId],
+    pub(crate) write_types: &'a [TypeId],
+}
+
+impl<'a> DynamicQueryTypes<'a> {
+    /// Creates a dynamic query that reads component types in `read_types`
+    /// and writes component types in `write_types`.
+    pub fn new(read_types: &'a [TypeId], write_types: &'a [TypeId]) -> Self {
+        Self {
+            read_types,
+            write_types,
+        }
+    }
+
+    /// Gets the component types to be read by this query.
+    pub fn read_types(&self) -> &[TypeId] {
+        self.read_types
+    }
+
+    /// Gets the component types to be accessed mutably by this query.
+    pub fn write_types(&self) -> &[TypeId] {
+        self.write_types
+    }
+}
+
+/// A pointer to a slice of component data from a [`DynamicQuery`].
+pub struct ComponentSlice {
+    typ: TypeId,
+    ptr: NonNull<u8>,
+    len: usize,
+    component_layout: Layout,
+}
+
+impl ComponentSlice {
+    /// Returns the number of components in this slice.
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns the layout of a single component in this slice.
+    pub fn component_layout(&self) -> Layout {
+        self.component_layout
+    }
+
+    /// Returns the number of bytes in the component slice.
+    pub fn len_in_bytes(&self) -> usize {
+        self.len() * self.component_layout().size()
+    }
+
+    /// Returns a pointer to the raw component data.
+    pub fn ptr(&self) -> NonNull<u8> {
+        self.ptr
+    }
+
+    /// Casts this component slice to a slice of type `T`.
+    ///
+    /// # Panics
+    /// Panics if `T` is not the type of the components in this slice.
+    pub fn as_slice<T: 'static>(&self) -> &[T] {
```

Comment:
This is unsound because `&[T]` may outlive the query (and indeed the entire `World`. `ComponentSlice` should have a lifetime parameter bounded by the lifetime of the query, which should be propagated to this return value.